### PR TITLE
[Woo POS] Centring of views

### LINKED ISSUE #13701: [Woo POS] Design Tweaks: Centering of views
1. Product error and empty views are not properly centered

| Design | App |
|--------|-----|
| image | image |

| Design                        | App                           |
|-------------------------------|-------------------------------|
| image | Simulator Screenshot - iPad Air 11-inch (M2) - 2024-08-22 at 14 29 07 |

2. Cart placeholder is not centered according to designs

| Design                        | App                           |
|-------------------------------|-------------------------------|
| Image | Image |

3. Payment views

| Design                        | App                           |
|-------------------------------|-------------------------------|
| Image | Image |

### PR DESCRIPTION
**Pull request by joshheald**
Closes: #13701 

## Description

This PR fixes some issues with incorrect vertical positioning of certain views in the POS, compared to designs.

1. Product error and empty views are not properly centered

| Design | App (old) | App (new) |
|--------|----------|------------|
| image | image | products load error centring |
| image | Simulator Screenshot - iPad Air 11-inch (M2) - 2024-08-22 at 14 29 07 | products empty state centring |

N.B. I also put the correct image in for the magnifier. We already have a very similar image in the main app, but it's a darker grey for POS, and since it's two toned I can't just use it as a template.

2. Cart placeholder 

| Design | App (old) | App (new) |
|--------|----------|------------|
| Image | Image | cart empty state centring |

3. Payment views

| Design | App (old) | App (new) |
|--------|----------|------------|
| Image | Image | payment error |

## Steps to reproduce

Set up Proxyman or Charles to have a breakpoint on the response to `https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/*/rest-api/?json=true&path=/wc/v3/products*` and proxy your device's connection

### Products load error

1. Launch the app and navigate to POS
2. Abort the first request (and any retries)
3. Observe that the payments error state matches designs – i.e. centred vertically on the screen, not between other views.

### Products empty state

1. Continue from above
2. Tap Retry
3. Delete the content of the `data` array in the response
4. Execute the edited response
5. Observe that the empty state matches designs – i.e. centred vertically on the screen, not between the cart and foot of the screen.

### Cart empty state

Turn off your breakpoints

1. Exit POS, and enter again
2. Observe that the empty cart shows, and the positioning matches designs – i.e. the text is centred vertically on the screen, and the image is ~40 points above it.

N.B. Strictly, the design is centred between the header and the invisible button for Check out, but that is essentially the same as centred vertically on screen. Also, the way I've done it means I can't be totally precise for the space between the image and the text, but this was the most sensible way I could find to centre the text and then position the image relative to it.

### Fullscreen payments errors

1. Add a product to your cart with an error amount, e.g. ending in .01
2. Connect your reader
3. Go through to checkout
4. Tap your card when prompted
5. Observe that the error is shown with 120px padding from the top of the screen, not vertically centred.

## Testing information

It's good to test the surrounding views especially for payments. I checked all the states I could think of, and also tested portrait mode and split view, such as they are.

No unit tests for this UI work.

## Screenshots

https://github.com/user-attachments/assets/70291cda-14df-4f44-ab62-7327d6e3f8be

---

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/221">20.2</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/POS/Presentation/CartView.swift [hunk 1]
```diff
@@ -224,6 +217,29 @@ private extension CartView {
             }
         }
     }
+
+    var cartEmptyView: some View {
+        VStack {
+            Spacer()
+            // By designs, the text should be vertically centred with the image 40px above it.
+            // SwiftUI doesn't allow us to absolutely pin a view to the centre then position other views relative to it
+            // Instead, we can centre the text, and then put the image in an offset overlay.
```
Comment (staskus) [inline WooCommerce/Classes/POS/Presentation/CartView.swift]: > // Instead, we can centre the text, and then put the image in an offset overlay.

That's a smart approach. Unfortunately, it doesn't quite support larger dynamic-type sizes.

image

File: WooCommerce/Classes/POS/Presentation/TotalsView.swift [hunk 2]
```diff
@@ -267,15 +267,31 @@ private extension TotalsView {
             topPadding: 40,
             bottomPadding: 40
         )
+
+        static let topAligned = CardReaderViewLayout(
```
Comment (staskus) [inline WooCommerce/Classes/POS/Presentation/TotalsView.swift:271]: I admit seeing a non-centered payment error view looked weird to me at first 😀.

Figma comment mentioned that it was done to make sure that there's always space for "No internet connection" banner.

Either way, happy that `CardReaderViewLayout` has proven to be useful in one more case.

**Review by staskus — APPROVED** [review]
✅ Thanks! Solid work.

I tested all the changed views: product error, product empty, cart empty, payment error in different configurations.

The only issue I found is with Cart + Dynamic Type.

File: WooCommerce/Classes/POS/Presentation/CartView.swift (hunk 1 above)
Comment (staskus) [reply WooCommerce/Classes/POS/Presentation/CartView.swift]: Using overlay looks to be a good option but it doesn't give us that control of layout. This is one of the things that would be easy to do with `UIKit` constraints but seems to be much harder to do with `SwiftUI` 🤔 I'm interested how to achieve it nicely as well. I'll let you know if I encounter a good solution.

**Comment by staskus** [comment]
Another question.

Does the payment view usually chop like that for you or does it only appear in the video, @joshheald ?

image

File: WooCommerce/Classes/POS/Presentation/CartView.swift [hunk 3]
```diff
@@ -224,6 +217,29 @@ private extension CartView {
...
+            Text(Localization.addItemsToCartHint)
+                .font(Constants.secondaryFont)
+                .lineSpacing(Constants.cartEmptyViewTextLineHeight)
+                .foregroundColor(Color.posTertiaryText)
+                .multilineTextAlignment(.center)
+                .overlay(alignment: .center) {
```
Comment (staskus) [inline WooCommerce/Classes/POS/Presentation/CartView.swift]: Turns out dynamic type issue is easy solvable by aligning this overlay on `top`.

```suggestion
                .overlay(alignment: .top) {
```

File: WooCommerce/Classes/POS/Presentation/TotalsView.swift (hunk 2 above)
Comment (joshheald) [reply WooCommerce/Classes/POS/Presentation/TotalsView.swift:271]: Yeah I was dreading this one when I saw how the code was written, then really happy to come across the Layout ❤️

**Comment by joshheald** [comment]
> Another question.
> 
> Does the payment view usually chop like that for you or does it only appear in the video, @joshheald ?

Thanks for the reminder – yes, I noticed that and wanted to check trunk to see whether it's introduced here, somehow. I don't see how it would be, but hey, transitions are fickle.

File: WooCommerce/Classes/POS/Presentation/CartView.swift (hunk 1 above)
Comment (joshheald) [reply WooCommerce/Classes/POS/Presentation/CartView.swift]: Dangit. Thanks for looking further and finding a fix, I'll put that in

**Comment by joshheald** [comment]
@staskus the cut between `Remove card` and `Processing` happens for me at 3845628, the commit on trunk that I branched from, and the one where I merged the processing animation. So I'm going to merge this PR, because it's unrelated to these changes.

It still happens on latest trunk too. I'll take a look at it separately, it's probably from my last PR.

**Pull request merged**